### PR TITLE
Never consider raw pointer casts to be trival

### LINKED ISSUE #113257: Raw pointer cast triggers "lifetime may not live long enough"
I tried this code:

```rust
struct Foo<'a>(&'a ());

fn foo(v: *const Foo<'_>) -> *const Foo<'static> {
    v as *const Foo<'static>
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=f2f7b3d9c228a4d9d77a711a5cae6691).

I expected to see this happen: it compiles, since raw pointer casts should be allowed even for completely different types.

Instead, this happened: it fails with:
```
error: lifetime may not live long enough
 --> src/lib.rs:4:5
  |
3 | fn foo(v: *const Foo<'_>) -> *const Foo<'static> {
  |        - has type `*const Foo<'1>`
4 |     v as *const Foo<'static>
  |     ^^^^^^^^^^^^^^^^^^^^^^^^ type annotation requires that `'1` must outlive `'static`
```
It looks like for some reason the compiler creates a lifetime obligation here.

Strangely, it works with `cast()`:
```rust
fn foo(v: *const Foo<'_>) -> *const Foo<'static> {
    v.cast::<Foo<'static>>()
}
```
[Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=3e808e171876b9cd2d0c7bb2a190b5a3).

### PR DESCRIPTION
**Pull request by Noratrieb**
HIR typeck tries to figure out which casts are trivial by doing them as
coercions and seeing whether this works. Since HIR typeck is oblivious
of lifetimes, this doesn't work for pointer casts that only change the
lifetime of the pointee, which are, as borrowck will tell you, not
trivial.

This change makes it so that raw pointer casts are never considered
trivial.

This also incidentally fixes the "trivial cast" lint false positive on
the same code. Unfortunately, "trivial cast" lints are now never emitted
on raw pointer casts, even if they truly are trivial. This could be
fixed by also doing the lint in borrowck for raw pointers specifically.

fixes #113257

### CODE REVIEW

File: compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs [hunk 1]
```diff
@@ -150,6 +150,7 @@ impl<'tcx> TypeErrCtxtExt<'tcx> for TypeErrCtxt<'_, 'tcx> {
         let trait_ref = trait_ref.skip_binder();
 
         let mut flags = vec![];
+        let mut display_flags = vec![];
```
Comment (Nilstrieb) [inline compiler/rustc_trait_selection/src/traits/error_reporting/on_unimplemented.rs:153]: oops that was a mistake

**Comment by saethlin** [comment]
We have a MIR opt that deletes redundant casts. Would it simplify things to always emit casts and lean on the opt?

**Comment by compiler-errors** [comment]
> We have a MIR opt that deletes redundant casts.

Yeah, perhaps we could do a perf run on that and see what it syas.

**Comment by Nilstrieb** [comment]
I looked a bit more into this. Saethlin's suggestion is not really possible. In MIR, all casts are "fully resolved", it's exactly known what kind of cast it is: https://doc.rust-lang.org/1.71.0/nightly-rustc/rustc_middle/mir/enum.CastKind.html. So we can't just create completely redundant casts like that (there is a MIR opt to remove redundant casts and it will nicely remove the pointer cast from #113257, but we can't just generate "a cast" because MIR has no notion of "a cast".

When exploring more about how casts work, I wrote down some stuff about them: https://hackmd.io/@Nilstrieb/S1eYpMXF3

So the question that remains for me is whether to do this pointer special casing (I am convinced that we need pointer special casing, because pointer casts _are_ special) should live in THIR building like my PR here or whether it should live in typeck.

Unless you have strong preferences, I would need to take a look at the code again to determine where it would best be placed.

File: tests/ui/cast/ptr-to-ptr-different-regions.rs [hunk 2]
```diff
@@ -0,0 +1,16 @@
+// check-pass
+
+// https://github.com/rust-lang/rust/issues/113257
+
+struct Foo<'a> { a: &'a () }
+
+fn extend_lifetime_very_very_safely<'a>(v: *const Foo<'a>) -> *const Foo<'static> {
+    // This should pass because raw pointer casts can do anything they want.
+    v as *const Foo<'static>
+}
```
Comment (RalfJung) [inline tests/ui/cast/ptr-to-ptr-different-regions.rs:10]: Could you add something like my example from https://github.com/rust-lang/rust/issues/114798 to this test, to make sure it also works for dyn trait pointer lifetime bounds?

**Review by lcnr — COMMENTED** [review]
hm, I think it would be better to change this in `CastCheck::check/try_coercion_cast` to not `set_coercion_cast` for pointer casts. Right now this PR would fix this for mir borrowck but it would still incorrectly emit the trivial cast lint.

Unsure how to handle this :thinking: never succeed in `try_coercion_cast` if we're casting raw pointers feels suboptimal but might be the best solution for now? I think the lint for raw pointers only makes sense after borrowck :shrug:

**Comment by Nilstrieb** [comment]
I don't have time to actually implement this right now but right now the solution I like the most is to add a new special `if` case to this branch here to always call `do_check` if we're doing a raw pointer cast. It feels wrong to have such a special case but I think we need a strong special case here.

https://github.com/rust-lang/rust/blob/4a0402cd805aa1f4dd8e2d96b038be664db293c3/compiler/rustc_hir_typeck/src/cast.rs#L729

**Comment by Nilstrieb** [comment]
I changed it a bit to now do the check in HIR typeck. I believe this is the best solution without rewriting how casting works everywhere (I haven't thought about what such a solution would look like), but I am not very confident.
@rustbot ready

**Comment by lcnr** [comment]
will take a look soon

r? @lcnr

File: compiler/rustc_hir_typeck/src/cast.rs [hunk 3]
```diff
@@ -660,9 +660,21 @@ impl<'a, 'tcx> CastCheck<'tcx> {
...
-                    debug!(" -> CoercionCast");
-                    fcx.typeck_results.borrow_mut().set_coercion_cast(self.expr.hir_id.local_id);
+                    if self.expr_ty.is_unsafe_ptr() && self.cast_ty.is_unsafe_ptr() {
+                        // When casting a raw pointer to another raw pointer, we cannot convert the cast into
+                        // a coercion because the pointee types might only differ in regions, which HIR typeck
+                        // cannot distinguish. This would cause us to erroneously discard a cast which will
+                        // lead to a borrowck error like #113257.
+                        // We still did a coercion above to unify inference variables for `ptr as _` casts.
+                        debug!(" -> PointerCast");
+                        // This does cause us to miss some trivial casts in the trival cast lint.
```
Comment (lcnr) [inline compiler/rustc_hir_typeck/src/cast.rs:670]:
```suggestion
                        // This does cause us to miss some trivial casts in the trival cast lint.
                        debug!(" -> PointerCast");
```

**Review by lcnr — COMMENTED** [review]
@Nilstrieb can you open an issue that we stop emitting the `trivial_casts` lint for pointers with an example?

---

going to nominate for @rust-lang/lang. I consider this to be a pure bugfix.

We currently do not generate any MIR for pointer casts which only transmute lifetimes. This happens as HIR typeck does not consider the source and target to have distinct types. This causes the following snippet to error:
```rust
struct Foo<'a>(&'a ());

fn foo<'a>(v: *const Foo<'a>) -> *const Foo<'static> {
    v as *const Foo<'static>
    //~^ ERROR lifetime may not live long enough
}
```
This can already be avoided by first casting to an unrelated type: 
```rust
struct Foo<'a>(&'a ());

fn foo<'a>(v: *const Foo<'a>) -> *const Foo<'static> {
    v as *const () as *const Foo<'static>
}
```
With this PR both snippets now compile. Unfortunately this means that we now don't emit the `trivial_casts` lint when casting raw pointers as we don't know whether the cast is trivial during HIR typeck:
```rust
#![warn(trivial_casts)]
struct Foo<'a>(&'a ());
fn foo<'a>(v: *const Foo<'a>) -> *const Foo<'a> {
    v as *const Foo<'a>
    //~^ WARN trivial cast: `*const Foo<'a>` as `*const **Foo<'a>`
}
```
We could check raw pointer casts during MIR borrowck and lint there, but that may not be worth it :shrug: I think it's fine to open an issue for it and merge this as is, as it does not seem too important to me.

**Comment by traviscross** [comment]
@lcnr: This was discussed in the T-lang meeting today.  T-lang agrees that this is a bug fix.

@rustbot labels -I-lang-nominated

**Comment by Nilstrieb** [comment]
I created an issue: #117187.

**Comment by steffahn** [comment]
You might be interested in

* #120217

which was caused or uncovered (or at least simplified to exploit, in case there was a different way before) by this PR.

**Pull request merged**